plugins/coreboot/fu-coreboot-fmap.c

```diff
@@ -0,0 +1,237 @@
+	fu_device_set_name (dev, name);
+	fu_device_set_vendor (dev, fu_device_get_vendor (parent));
+
+	sum = g_strdup_printf ("%s, partition '%s'\n", fu_device_get_summary (parent), fmap_area->name);
+	fu_device_set_summary (dev, sum);
+
+	fu_device_add_flag (dev, FWUPD_DEVICE_FLAG_INTERNAL);
+	fu_device_add_icon (dev, "computer");
+	inst_id = g_strdup_printf ("system-fimware-partition-%s\n", fmap_area->name);
+	fu_device_add_instance_id (dev, "system-fimware-partition");
+	fu_device_add_instance_id (dev, inst_id);
+	fu_device_add_parent_guid (dev, "main-system-firmware");
+
+	fu_device_set_metadata (dev, FU_DEVICE_METADATA_FLASHROM_DEVICE_KIND, "system-firmware");
+	fu_device_set_metadata (dev, FU_DEVICE_METADATA_FLASHROM_FMAP_NAME, (const char *)fmap_area->name);
+
+	if (!immutable)
+		fu_device_add_flag (dev, FWUPD_DEVICE_FLAG_UPDATABLE);
+	fu_device_set_firmware_size_max (dev, fmap_area->size);
+
+	// Fill in parent version as fallback
+	//fu_device_set_version (dev, fu_device_get_version (parent), FWUPD_VERSION_FORMAT_TRIPLET);
+
+	//FIXME: that only works on x86 with STRICT_DEV_MEM disabled
+	//The correct way would be to use libflashrom to read parts of the flash
+#if 0
+	// try to get version from CBFS in current partition
+	region = fu_plugin_coreboot_clone_region (plugin, error, fmap, fmap_area);
+	if (region) {
+		guint8 *file = fu_plugin_coreboot_get_raw_cbfs_file (region, fmap_area->size, "revision");
+		if (file) {
+			g_debug ("Found file 'revision' in CBFS@%s", fmap_area->name);
+			triplet = fu_plugin_coreboot_parse_revision_file ((const gchar *)file, error);
+			g_debug ("FMAP coreboot version %s", triplet);
+			if (triplet)
+				fu_device_set_version (dev, triplet, FWUPD_VERSION_FORMAT_TRIPLET);
+		}
+		g_free (region);
+	} else {
+		g_debug ("Failed to MMAP region %s", fmap_area->name);
+	}
+#endif
+	return dev;
+
+}
+
+/* Return a pointer to the corresponding fmap_area by name.
+ * Returns NULL if not found.
+ */
+static struct fmap_area *
+fu_plugin_coreboot_fmap_area_by_name (const struct fmap *fmap, const char *name)
+{
+	for (gint i = 0; i < fmap->nareas; i++) {
+		if (g_strcmp0 (name, (const char *)fmap->areas[i].name) == 0)
+			return &fmap->areas[i];
+	}
+	return NULL;
+}
+
+/* Add FMAP partitions based on 'known' names */
+gboolean
+fu_plugin_coreboot_add_fmap_devices (FuPlugin *plugin, GError **error, FuDevice *parent, const struct fmap *fmap)
+{
+	static const char *regions[] = {
+		// VBOOT enabled devices have up to two R/W partitions:
+		"RW_SECTION_A",
+		"RW_SECTION_B",
+		// some VBOOT enabled devices have one RO section:
+		"RO_SECTION",
+		// Autogenerated FMAP on x86 has one region:
+		"BIOS",
+	};
+	struct fmap_area *area;
+	gint vboot_rw_partition_cnt = 0;
+	FuDevice *dev;
+
+	if (fu_plugin_coreboot_fmap_area_by_name (fmap, "RW_SECTION_A"))
+		vboot_rw_partition_cnt++;
+	if (fu_plugin_coreboot_fmap_area_by_name (fmap, "RW_SECTION_B"))
+                vboot_rw_partition_cnt++;
```

Comment:
This seems like dead code.  `vboot_rw_partition_cnt` isn't actually used in the function for something useful.  If you do think it's needed, then I think it should be moved into the for loop below to count while you're iterating.

Reply:
I still need to figure out which partition to update and the priority, depending on the active partition and partition count.